go/marshal/decode.go

```diff
@@ -0,0 +1,227 @@
+// Copyright 2016 Attic Labs, Inc. All rights reserved.
+// Licensed under the Apache License, version 2.0:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+package marshal
+
+import (
+	"fmt"
+	"reflect"
+	"runtime"
+	"sync"
+
+	"github.com/attic-labs/noms/go/types"
+)
+
+// Unmarshal converts a Noms value into a Go value. It decodes v and stores the result
+// in the value pointed to by out.
+//
+// Unmarshal uses the inverse of the encodings that Marshal uses with the following additional rules:
+//
+// To unmarshal a Noms struct into a Go struct, Unmarshal matches incoming object
+// fields to the fields used by Marshal (either the struct field name or its tag).
+// Unmarshal will only set exported fields of the struct.
+//
+// If a Noms value is not appropriate for a given target type, or if a Noms number overflows the target type, Unmarshal returns a UnmarshalTypeMismatchError or OverflowError respectively
+//
+func Unmarshal(v types.Value, out interface{}) (err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			if _, ok := r.(runtime.Error); ok {
+				panic(r)
+			}
+			if s, ok := r.(string); ok {
+				panic(s)
```

Comment:
I'm not sure. I followed the pattern in encoding/json. I'm going to go through these and see if I can reduce the number of different errors I use and try to figure out the `runtime.Error` case too,